java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -83,48 +119,67 @@ class DirectStoreMuxer<Data : CrdtData, Op : CrdtOperation, T>(
     }
 
     /**
-     * Sends the provided [ProxyMessage] to the store backing the provided [referenceId].
+     * Sends the provided [ProxyMessage] to the store backing the provided [muxId].
      *
-     * A new store will be created for the [referenceId], if necessary.
+     * A new store will be created for the [muxId], if necessary.
      */
     suspend fun onProxyMessage(
-        message: ProxyMessage<Data, Op, T>,
-        referenceId: String
+        message: ProxyMessage<Data, Op, T>
     ): Boolean {
-        val (id, store) = store(referenceId)
-        val deMuxedMessage: ProxyMessage<Data, Op, T> = when (message) {
-            is SyncRequest -> SyncRequest(id)
-            is ModelUpdate -> ModelUpdate(message.model, id)
-            is Operations -> if (message.operations.isNotEmpty()) {
-                Operations(message.operations, id)
-            } else return true
+        val muxId = requireNotNull(message.muxId) {
+            "messages sent to Direct Store Muxer must have a muxId"
         }
-        return store.onProxyMessage(deMuxedMessage)
+        val (idSet, store) = store(muxId)
+
+        if (!idSet.contains(message.id)) registerToStore(message.id!!, muxId)
+
+        return store.onProxyMessage(message)
     }
 
-    /* internal */ suspend fun setupStore(referenceId: String): StoreRecord<Data, Op, T> {
+    /* internal */ suspend fun setupStore(muxId: String): StoreRecord<Data, Op, T> {
         val store = DirectStore.create<Data, Op, T>(
             StoreOptions(
-                storageKey = storageKey.childKeyWithComponent(referenceId),
+                storageKey = storageKey.childKeyWithComponent(muxId),
                 type = backingType
             )
         )
 
-        val id = store.on(callbackFactory(referenceId))
+        // Return a new Record with initially empty idSet
+        return StoreRecord(mutableSetOf<Int>(), store)
+    }
 
-        // Return a new Record and add it to our local stores, keyed by muxId.
-        return StoreRecord(id, store)
+    private suspend fun registerToStore(
+        id: Int,
+        muxId: String
+    ) {
+        // The Direct Store Muxer can register several callbacks to the same store.
+        // When the Direct Store Muxer receives a message from a Direct Store, it utilizes the
+        // message id (determined by which callback the Direct Store invoked) to determine which
+        // observer to redirect the message to.
+        val (idSet, store) = store(muxId)
+        storeMutex.withLock {
+            if (!idSet.contains(id)) {
```

Comment:
I also would propose moving this check to outside the mutex to reduce contention? ie
val (idSet, store) = store(muxId)
if (!idSet.contains(id)) {
     storeMutex.withLock {
      ...

Reply:
Moving the check to outside the mutex causes the `directStoreMuxerNoRace` test to crash consistently.